File: android/app/src/main/kotlin/com/sipeed/zeroclaw/StartupTrace.kt
package com.sipeed.zeroclaw

import android.os.SystemClock
import android.util.Log

object StartupTrace {
    private const val TAG = "StartupTrace"
    private val processStartElapsedMs = SystemClock.elapsedRealtime()

    fun mark(label: String, details: String? = null) {
        val elapsedMs = SystemClock.elapsedRealtime() - processStartElapsedMs
        val suffix = if (details.isNullOrBlank()) "" else " | $details"
        Log.i(TAG, "[$elapsedMs" + "ms] $label$suffix")
    }
}

File: lib/src/core/startup_trace.dart
import 'dart:developer' as developer;

import 'package:flutter/foundation.dart';
import 'package:flutter/widgets.dart';

class StartupTrace {
  StartupTrace._();

  static final Stopwatch _stopwatch = Stopwatch()..start();
  static final Set<String> _onceMarkers = <String>{};

  static void mark(String label, {String? details}) {
    final elapsedMs = _stopwatch.elapsedMilliseconds;
    final suffix = details == null || details.isEmpty ? '' : ' | $details';
    final message = '[StartupTrace][${elapsedMs}ms] $label$suffix';
    developer.log(message, name: 'StartupTrace');
    debugPrint(message);
  }

  static void markOnce(String label, {String? details}) {
    if (_onceMarkers.add(label)) {
      mark(label, details: details);
    }
  }

  static void attachFirstFrameTrace() {
    WidgetsBinding.instance.addPostFrameCallback((_) {
      markOnce('Flutter first frame rasterized');
    });
  }
}
